irksome/nystrom_stepper.py

```diff
@@ -0,0 +1,137 @@
+from .base_time_stepper import StageCoupledTimeStepper
+from .bcs import BCStageData, bc2space
+from .deriv import TimeDerivative
+from .tools import component_replace, replace, vecconst
+from firedrake import TestFunction, as_ufl
+import numpy
+from ufl import zero
+
+
+class NystromTableau:
```

Comment:
Should ButcherTableau inherit from this class?

Reply:
Maybe the other way (this inherits from BT)?  Not sure of the optimal design here.  Every Nystrom tableau has A, b, c but not every Butcher tableau has Abar, bbar .  Every regular Butcher tableau can be promoted by `Abar = A @ A` and `bbar = A.T @ b`.  However, supporting queries (is this method explicit/implicit etc?) has to be overloaded in the Nystrom case since you have two matrices to check instead of one.

Reply:
I'm with @rckirby here - I think Nystrom inherits from Butcher, since it's Butcher + more stuff.  

I don't know that there's much to be done about it, but I notice that we're starting to propagate slightly ambiguous additional data in these Tableau.  Nystrom has `Abar` and `bbar`, DIRK-IMEX have `A_hat` and `b_hat`.  Regular Butcher Tableau have `btilde` for possibly adaptivity.  It's not our fault, but collisions here appear inevitable.  For example, the notation in Hairer-Wanner-Lubich for Partitioned schemes have `A_hat` and `b_hat` again.  Do we care?  Is there a good way to keep track of these?

Reply:
Structurally, we have common fields between Nystrom & partitioned, but they do different things.  I think there is a high potential for confusion if we try to make them share code?

Inheriting Nystrom from standard Butcher is mathematically a bit dubious -- there exist Nystrom methods that aren't derived from RK methods.  If a Nystrom tableau IS a Butcher tableau (in the sense of inheritance), that should guarantee one could use it for a 1st order problem.  Which isn't true.

Reply:
I'm convinced that we should have constructors for Butcher -> Nystrom and (eventually) partitioned -> Nystrom, but not inheritance.

That doesn't remove the need to discuss our naming convention problem as we graft additional data on to a standard Butcher Tableau.  We can push that down the road (and I'm happy to do so), but we are starting to run into ambiguity with our `A_hat`, `Abar`, etc., naming.